// box-node-sdk/lib-esm/schemas/metadataTaxonomyNode.js
import { serializeMetadataTaxonomyAncestor } from './metadataTaxonomyAncestor.js';
import { deserializeMetadataTaxonomyAncestor } from './metadataTaxonomyAncestor.js';
import { BoxSdkError } from '../box/errors.js';
import { sdIsNumber } from '../serialization/json.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsList } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export function serializeMetadataTaxonomyNode(val) {
    return {
        ['id']: val.id,
        ['displayName']: val.displayName,
        ['level']: val.level,
        ['parentId']: val.parentId,
        ['nodePath']: val.nodePath == void 0
            ? val.nodePath
            : val.nodePath.map(function (item) {
                return item;
            }),
        ['ancestors']: val.ancestors == void 0
            ? val.ancestors
            : val.ancestors.map(function (item) {
                return serializeMetadataTaxonomyAncestor(item);
            }),
    };
}
export function deserializeMetadataTaxonomyNode(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "MetadataTaxonomyNode"',
        });
    }
    if (val.id == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "id" of type "MetadataTaxonomyNode" to be defined',
        });
    }
    if (!sdIsString(val.id)) {
        throw new BoxSdkError({
            message: 'Expecting string for "id" of type "MetadataTaxonomyNode"',
        });
    }
    const id = val.id;
    if (val.displayName == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "displayName" of type "MetadataTaxonomyNode" to be defined',
        });
    }
    if (!sdIsString(val.displayName)) {
        throw new BoxSdkError({
            message: 'Expecting string for "displayName" of type "MetadataTaxonomyNode"',
        });
    }
    const displayName = val.displayName;
    if (val.level == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "level" of type "MetadataTaxonomyNode" to be defined',
        });
    }
    if (!sdIsNumber(val.level)) {
        throw new BoxSdkError({
            message: 'Expecting number for "level" of type "MetadataTaxonomyNode"',
        });
    }
    const level = val.level;
    if (!(val.parentId == void 0) && !sdIsString(val.parentId)) {
        throw new BoxSdkError({
            message: 'Expecting string for "parentId" of type "MetadataTaxonomyNode"',
        });
    }
    const parentId = val.parentId == void 0 ? void 0 : val.parentId;
    if (!(val.nodePath == void 0) && !sdIsList(val.nodePath)) {
        throw new BoxSdkError({
            message: 'Expecting array for "nodePath" of type "MetadataTaxonomyNode"',
        });
    }
    const nodePath = val.nodePath == void 0
        ? void 0
        : sdIsList(val.nodePath)
            ? val.nodePath.map(function (itm) {
                if (!sdIsString(itm)) {
                    throw new BoxSdkError({
                        message: 'Expecting string for "MetadataTaxonomyNode"',
                    });
                }
                return itm;
            })
            : [];
    if (!(val.ancestors == void 0) && !sdIsList(val.ancestors)) {
        throw new BoxSdkError({
            message: 'Expecting array for "ancestors" of type "MetadataTaxonomyNode"',
        });
    }
    const ancestors = val.ancestors == void 0
        ? void 0
        : sdIsList(val.ancestors)
            ? val.ancestors.map(function (itm) {
                return deserializeMetadataTaxonomyAncestor(itm);
            })
            : [];
    return {
        id: id,
        displayName: displayName,
        level: level,
        parentId: parentId,
        nodePath: nodePath,
        ancestors: ancestors,
    };
}
