// box-node-sdk/lib-esm/schemas/metadataTaxonomy.js
import { serializeMetadataTaxonomyLevel } from './metadataTaxonomyLevel.js';
import { deserializeMetadataTaxonomyLevel } from './metadataTaxonomyLevel.js';
import { BoxSdkError } from '../box/errors.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsList } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export function serializeMetadataTaxonomy(val) {
    return {
        ['id']: val.id,
        ['key']: val.key,
        ['displayName']: val.displayName,
        ['namespace']: val.namespace,
        ['levels']: val.levels == void 0
            ? val.levels
            : val.levels.map(function (item) {
                return serializeMetadataTaxonomyLevel(item);
            }),
    };
}
export function deserializeMetadataTaxonomy(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "MetadataTaxonomy"',
        });
    }
    if (val.id == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "id" of type "MetadataTaxonomy" to be defined',
        });
    }
    if (!sdIsString(val.id)) {
        throw new BoxSdkError({
            message: 'Expecting string for "id" of type "MetadataTaxonomy"',
        });
    }
    const id = val.id;
    if (!(val.key == void 0) && !sdIsString(val.key)) {
        throw new BoxSdkError({
            message: 'Expecting string for "key" of type "MetadataTaxonomy"',
        });
    }
    const key = val.key == void 0 ? void 0 : val.key;
    if (val.displayName == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "displayName" of type "MetadataTaxonomy" to be defined',
        });
    }
    if (!sdIsString(val.displayName)) {
        throw new BoxSdkError({
            message: 'Expecting string for "displayName" of type "MetadataTaxonomy"',
        });
    }
    const displayName = val.displayName;
    if (val.namespace == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "namespace" of type "MetadataTaxonomy" to be defined',
        });
    }
    if (!sdIsString(val.namespace)) {
        throw new BoxSdkError({
            message: 'Expecting string for "namespace" of type "MetadataTaxonomy"',
        });
    }
    const namespace = val.namespace;
    if (!(val.levels == void 0) && !sdIsList(val.levels)) {
        throw new BoxSdkError({
            message: 'Expecting array for "levels" of type "MetadataTaxonomy"',
        });
    }
    const levels = val.levels == void 0
        ? void 0
        : sdIsList(val.levels)
            ? val.levels.map(function (itm) {
                return deserializeMetadataTaxonomyLevel(itm);
            })
            : [];
    return {
        id: id,
        key: key,
        displayName: displayName,
        namespace: namespace,
        levels: levels,
    };
}
