// box-node-sdk/lib-esm/schemas/metadataTaxonomyLevel.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsNumber } from '../serialization/json.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export function serializeMetadataTaxonomyLevel(val) {
    return {
        ['displayName']: val.displayName,
        ['description']: val.description,
        ['level']: val.level,
    };
}
export function deserializeMetadataTaxonomyLevel(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "MetadataTaxonomyLevel"',
        });
    }
    if (!(val.displayName == void 0) && !sdIsString(val.displayName)) {
        throw new BoxSdkError({
            message: 'Expecting string for "displayName" of type "MetadataTaxonomyLevel"',
        });
    }
    const displayName = val.displayName == void 0 ? void 0 : val.displayName;
    if (!(val.description == void 0) && !sdIsString(val.description)) {
        throw new BoxSdkError({
            message: 'Expecting string for "description" of type "MetadataTaxonomyLevel"',
        });
    }
    const description = val.description == void 0 ? void 0 : val.description;
    if (!(val.level == void 0) && !sdIsNumber(val.level)) {
        throw new BoxSdkError({
            message: 'Expecting number for "level" of type "MetadataTaxonomyLevel"',
        });
    }
    const level = val.level == void 0 ? void 0 : val.level;
    return {
        displayName: displayName,
        description: description,
        level: level,
    };
}
