// box-node-sdk/lib-esm/schemas/metadataTaxonomyAncestor.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsNumber } from '../serialization/json.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export function serializeMetadataTaxonomyAncestor(val) {
    return {
        ['id']: val.id,
        ['displayName']: val.displayName,
        ['level']: val.level,
    };
}
export function deserializeMetadataTaxonomyAncestor(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "MetadataTaxonomyAncestor"',
        });
    }
    if (!(val.id == void 0) && !sdIsString(val.id)) {
        throw new BoxSdkError({
            message: 'Expecting string for "id" of type "MetadataTaxonomyAncestor"',
        });
    }
    const id = val.id == void 0 ? void 0 : val.id;
    if (!(val.displayName == void 0) && !sdIsString(val.displayName)) {
        throw new BoxSdkError({
            message: 'Expecting string for "displayName" of type "MetadataTaxonomyAncestor"',
        });
    }
    const displayName = val.displayName == void 0 ? void 0 : val.displayName;
    if (!(val.level == void 0) && !sdIsNumber(val.level)) {
        throw new BoxSdkError({
            message: 'Expecting number for "level" of type "MetadataTaxonomyAncestor"',
        });
    }
    const level = val.level == void 0 ? void 0 : val.level;
    return {
        id: id,
        displayName: displayName,
        level: level,
    };
}
