// box-node-sdk/lib-esm/schemas/signRequestSignerInputSsnValidation.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export class SignRequestSignerInputSsnValidation {
    /**
     * Validates that the text input is a Social Security Number (SSN). */
    validationType = 'ssn';
    rawData;
    constructor(fields) {
        if (fields.validationType !== undefined) {
            this.validationType = fields.validationType;
        }
        if (fields.rawData !== undefined) {
            this.rawData = fields.rawData;
        }
    }
}
export function serializeSignRequestSignerInputSsnValidationValidationTypeField(val) {
    return val;
}
export function deserializeSignRequestSignerInputSsnValidationValidationTypeField(val) {
    if (val == 'ssn') {
        return val;
    }
    if (sdIsString(val)) {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize SignRequestSignerInputSsnValidationValidationTypeField",
    });
}
export function serializeSignRequestSignerInputSsnValidation(val) {
    return {
        ['validation_type']: serializeSignRequestSignerInputSsnValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputSsnValidation(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputSsnValidation"',
        });
    }
    if (val.validation_type == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "validation_type" of type "SignRequestSignerInputSsnValidation" to be defined',
        });
    }
    const validationType = deserializeSignRequestSignerInputSsnValidationValidationTypeField(val.validation_type);
    return {
        validationType: validationType,
    };
}
export function serializeSignRequestSignerInputSsnValidationInput(val) {
    return {
        ['validationType']: val.validationType == void 0
            ? val.validationType
            : serializeSignRequestSignerInputSsnValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputSsnValidationInput(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputSsnValidationInput"',
        });
    }
    const validationType = val.validationType == void 0
        ? void 0
        : deserializeSignRequestSignerInputSsnValidationValidationTypeField(val.validationType);
    return {
        validationType: validationType,
    };
}
