// box-node-sdk/lib-esm/schemas/signRequestSignerInputDateUsValidation.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export function serializeSignRequestSignerInputDateUsValidationValidationTypeField(val) {
    return val;
}
export function deserializeSignRequestSignerInputDateUsValidationValidationTypeField(val) {
    if (val == 'date_us') {
        return val;
    }
    if (sdIsString(val)) {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize SignRequestSignerInputDateUsValidationValidationTypeField",
    });
}
export function serializeSignRequestSignerInputDateUsValidation(val) {
    return {
        ['validation_type']: val.validationType == void 0
            ? val.validationType
            : serializeSignRequestSignerInputDateUsValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputDateUsValidation(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputDateUsValidation"',
        });
    }
    const validationType = val.validation_type == void 0
        ? void 0
        : deserializeSignRequestSignerInputDateUsValidationValidationTypeField(val.validation_type);
    return {
        validationType: validationType,
    };
}
