// box-node-sdk/lib-esm/schemas/signRequestSignerInputNumberWithPeriodValidation.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export function serializeSignRequestSignerInputNumberWithPeriodValidationValidationTypeField(val) {
    return val;
}
export function deserializeSignRequestSignerInputNumberWithPeriodValidationValidationTypeField(val) {
    if (val == 'number_with_period') {
        return val;
    }
    if (sdIsString(val)) {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize SignRequestSignerInputNumberWithPeriodValidationValidationTypeField",
    });
}
export function serializeSignRequestSignerInputNumberWithPeriodValidation(val) {
    return {
        ['validation_type']: val.validationType == void 0
            ? val.validationType
            : serializeSignRequestSignerInputNumberWithPeriodValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputNumberWithPeriodValidation(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputNumberWithPeriodValidation"',
        });
    }
    const validationType = val.validation_type == void 0
        ? void 0
        : deserializeSignRequestSignerInputNumberWithPeriodValidationValidationTypeField(val.validation_type);
    return {
        validationType: validationType,
    };
}
