// box-node-sdk/lib-esm/schemas/signRequestSignerInputDateAsiaValidation.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export function serializeSignRequestSignerInputDateAsiaValidationValidationTypeField(val) {
    return val;
}
export function deserializeSignRequestSignerInputDateAsiaValidationValidationTypeField(val) {
    if (val == 'date_asia') {
        return val;
    }
    if (sdIsString(val)) {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize SignRequestSignerInputDateAsiaValidationValidationTypeField",
    });
}
export function serializeSignRequestSignerInputDateAsiaValidation(val) {
    return {
        ['validation_type']: val.validationType == void 0
            ? val.validationType
            : serializeSignRequestSignerInputDateAsiaValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputDateAsiaValidation(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputDateAsiaValidation"',
        });
    }
    const validationType = val.validation_type == void 0
        ? void 0
        : deserializeSignRequestSignerInputDateAsiaValidationValidationTypeField(val.validation_type);
    return {
        validationType: validationType,
    };
}
