// box-node-sdk/lib-esm/schemas/signRequestSignerInputDateIsoValidation.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export function serializeSignRequestSignerInputDateIsoValidationValidationTypeField(val) {
    return val;
}
export function deserializeSignRequestSignerInputDateIsoValidationValidationTypeField(val) {
    if (val == 'date_iso') {
        return val;
    }
    if (sdIsString(val)) {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize SignRequestSignerInputDateIsoValidationValidationTypeField",
    });
}
export function serializeSignRequestSignerInputDateIsoValidation(val) {
    return {
        ['validation_type']: val.validationType == void 0
            ? val.validationType
            : serializeSignRequestSignerInputDateIsoValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputDateIsoValidation(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputDateIsoValidation"',
        });
    }
    const validationType = val.validation_type == void 0
        ? void 0
        : deserializeSignRequestSignerInputDateIsoValidationValidationTypeField(val.validation_type);
    return {
        validationType: validationType,
    };
}
