// box-node-sdk/lib-esm/schemas/signRequestSignerInputZip4Validation.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export class SignRequestSignerInputZip4Validation {
    /**
     * Validates that the text input is a ZIP+4 code. */
    validationType = 'zip_4';
    rawData;
    constructor(fields) {
        if (fields.validationType !== undefined) {
            this.validationType = fields.validationType;
        }
        if (fields.rawData !== undefined) {
            this.rawData = fields.rawData;
        }
    }
}
export function serializeSignRequestSignerInputZip4ValidationValidationTypeField(val) {
    return val;
}
export function deserializeSignRequestSignerInputZip4ValidationValidationTypeField(val) {
    if (val == 'zip_4') {
        return val;
    }
    if (sdIsString(val)) {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize SignRequestSignerInputZip4ValidationValidationTypeField",
    });
}
export function serializeSignRequestSignerInputZip4Validation(val) {
    return {
        ['validation_type']: serializeSignRequestSignerInputZip4ValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputZip4Validation(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputZip4Validation"',
        });
    }
    if (val.validation_type == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "validation_type" of type "SignRequestSignerInputZip4Validation" to be defined',
        });
    }
    const validationType = deserializeSignRequestSignerInputZip4ValidationValidationTypeField(val.validation_type);
    return {
        validationType: validationType,
    };
}
export function serializeSignRequestSignerInputZip4ValidationInput(val) {
    return {
        ['validationType']: val.validationType == void 0
            ? val.validationType
            : serializeSignRequestSignerInputZip4ValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputZip4ValidationInput(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputZip4ValidationInput"',
        });
    }
    const validationType = val.validationType == void 0
        ? void 0
        : deserializeSignRequestSignerInputZip4ValidationValidationTypeField(val.validationType);
    return {
        validationType: validationType,
    };
}
