// box-node-sdk/lib-esm/schemas/signRequestSignerInputEmailValidation.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export class SignRequestSignerInputEmailValidation {
    /**
     * Validates that the text input is an email address. */
    validationType = 'email';
    rawData;
    constructor(fields) {
        if (fields.validationType !== undefined) {
            this.validationType = fields.validationType;
        }
        if (fields.rawData !== undefined) {
            this.rawData = fields.rawData;
        }
    }
}
export function serializeSignRequestSignerInputEmailValidationValidationTypeField(val) {
    return val;
}
export function deserializeSignRequestSignerInputEmailValidationValidationTypeField(val) {
    if (val == 'email') {
        return val;
    }
    if (sdIsString(val)) {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize SignRequestSignerInputEmailValidationValidationTypeField",
    });
}
export function serializeSignRequestSignerInputEmailValidation(val) {
    return {
        ['validation_type']: serializeSignRequestSignerInputEmailValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputEmailValidation(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputEmailValidation"',
        });
    }
    if (val.validation_type == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "validation_type" of type "SignRequestSignerInputEmailValidation" to be defined',
        });
    }
    const validationType = deserializeSignRequestSignerInputEmailValidationValidationTypeField(val.validation_type);
    return {
        validationType: validationType,
    };
}
export function serializeSignRequestSignerInputEmailValidationInput(val) {
    return {
        ['validationType']: val.validationType == void 0
            ? val.validationType
            : serializeSignRequestSignerInputEmailValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputEmailValidationInput(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputEmailValidationInput"',
        });
    }
    const validationType = val.validationType == void 0
        ? void 0
        : deserializeSignRequestSignerInputEmailValidationValidationTypeField(val.validationType);
    return {
        validationType: validationType,
    };
}
