// box-node-sdk/lib-esm/schemas/signRequestSignerInputNumberWithCommaValidation.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export function serializeSignRequestSignerInputNumberWithCommaValidationValidationTypeField(val) {
    return val;
}
export function deserializeSignRequestSignerInputNumberWithCommaValidationValidationTypeField(val) {
    if (val == 'number_with_comma') {
        return val;
    }
    if (sdIsString(val)) {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize SignRequestSignerInputNumberWithCommaValidationValidationTypeField",
    });
}
export function serializeSignRequestSignerInputNumberWithCommaValidation(val) {
    return {
        ['validation_type']: val.validationType == void 0
            ? val.validationType
            : serializeSignRequestSignerInputNumberWithCommaValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputNumberWithCommaValidation(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputNumberWithCommaValidation"',
        });
    }
    const validationType = val.validation_type == void 0
        ? void 0
        : deserializeSignRequestSignerInputNumberWithCommaValidationValidationTypeField(val.validation_type);
    return {
        validationType: validationType,
    };
}
