// box-node-sdk/lib-esm/schemas/signRequestSignerInputZipValidation.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export class SignRequestSignerInputZipValidation {
    /**
     * Validates that the text input is a ZIP code. */
    validationType = 'zip';
    rawData;
    constructor(fields) {
        if (fields.validationType !== undefined) {
            this.validationType = fields.validationType;
        }
        if (fields.rawData !== undefined) {
            this.rawData = fields.rawData;
        }
    }
}
export function serializeSignRequestSignerInputZipValidationValidationTypeField(val) {
    return val;
}
export function deserializeSignRequestSignerInputZipValidationValidationTypeField(val) {
    if (val == 'zip') {
        return val;
    }
    if (sdIsString(val)) {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize SignRequestSignerInputZipValidationValidationTypeField",
    });
}
export function serializeSignRequestSignerInputZipValidation(val) {
    return {
        ['validation_type']: serializeSignRequestSignerInputZipValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputZipValidation(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputZipValidation"',
        });
    }
    if (val.validation_type == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "validation_type" of type "SignRequestSignerInputZipValidation" to be defined',
        });
    }
    const validationType = deserializeSignRequestSignerInputZipValidationValidationTypeField(val.validation_type);
    return {
        validationType: validationType,
    };
}
export function serializeSignRequestSignerInputZipValidationInput(val) {
    return {
        ['validationType']: val.validationType == void 0
            ? val.validationType
            : serializeSignRequestSignerInputZipValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputZipValidationInput(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputZipValidationInput"',
        });
    }
    const validationType = val.validationType == void 0
        ? void 0
        : deserializeSignRequestSignerInputZipValidationValidationTypeField(val.validationType);
    return {
        validationType: validationType,
    };
}
