// box-node-sdk/lib-esm/schemas/signRequestSignerInputDateEuValidation.js
import { BoxSdkError } from '../box/errors.js';
import { sdIsString } from '../serialization/json.js';
import { sdIsMap } from '../serialization/json.js';
export function serializeSignRequestSignerInputDateEuValidationValidationTypeField(val) {
    return val;
}
export function deserializeSignRequestSignerInputDateEuValidationValidationTypeField(val) {
    if (val == 'date_eu') {
        return val;
    }
    if (sdIsString(val)) {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize SignRequestSignerInputDateEuValidationValidationTypeField",
    });
}
export function serializeSignRequestSignerInputDateEuValidation(val) {
    return {
        ['validation_type']: val.validationType == void 0
            ? val.validationType
            : serializeSignRequestSignerInputDateEuValidationValidationTypeField(val.validationType),
    };
}
export function deserializeSignRequestSignerInputDateEuValidation(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputDateEuValidation"',
        });
    }
    const validationType = val.validation_type == void 0
        ? void 0
        : deserializeSignRequestSignerInputDateEuValidationValidationTypeField(val.validation_type);
    return {
        validationType: validationType,
    };
}
