// box-node-sdk/lib-esm/schemas/v2025R0/enterpriseConfigurationV2025R0.js
import { serializeEnterpriseConfigurationSecurityV2025R0 } from './enterpriseConfigurationSecurityV2025R0.js';
import { deserializeEnterpriseConfigurationSecurityV2025R0 } from './enterpriseConfigurationSecurityV2025R0.js';
import { serializeEnterpriseConfigurationContentAndSharingV2025R0 } from './enterpriseConfigurationContentAndSharingV2025R0.js';
import { deserializeEnterpriseConfigurationContentAndSharingV2025R0 } from './enterpriseConfigurationContentAndSharingV2025R0.js';
import { serializeEnterpriseConfigurationUserSettingsV2025R0 } from './enterpriseConfigurationUserSettingsV2025R0.js';
import { deserializeEnterpriseConfigurationUserSettingsV2025R0 } from './enterpriseConfigurationUserSettingsV2025R0.js';
import { serializeEnterpriseConfigurationShieldV2025R0 } from './enterpriseConfigurationShieldV2025R0.js';
import { deserializeEnterpriseConfigurationShieldV2025R0 } from './enterpriseConfigurationShieldV2025R0.js';
import { BoxSdkError } from '../../box/errors.js';
import { sdIsString } from '../../serialization/json.js';
import { sdIsMap } from '../../serialization/json.js';
export function serializeEnterpriseConfigurationV2025R0TypeField(val) {
    return val;
}
export function deserializeEnterpriseConfigurationV2025R0TypeField(val) {
    if (val == 'enterprise_configuration') {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize EnterpriseConfigurationV2025R0TypeField",
    });
}
export function serializeEnterpriseConfigurationV2025R0(val) {
    return {
        ['id']: val.id,
        ['type']: val.type == void 0
            ? val.type
            : serializeEnterpriseConfigurationV2025R0TypeField(val.type),
        ['security']: val.security == void 0
            ? val.security
            : serializeEnterpriseConfigurationSecurityV2025R0(val.security),
        ['content_and_sharing']: val.contentAndSharing == void 0
            ? val.contentAndSharing
            : serializeEnterpriseConfigurationContentAndSharingV2025R0(val.contentAndSharing),
        ['user_settings']: val.userSettings == void 0
            ? val.userSettings
            : serializeEnterpriseConfigurationUserSettingsV2025R0(val.userSettings),
        ['shield']: val.shield == void 0
            ? val.shield
            : serializeEnterpriseConfigurationShieldV2025R0(val.shield),
    };
}
export function deserializeEnterpriseConfigurationV2025R0(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "EnterpriseConfigurationV2025R0"',
        });
    }
    if (!(val.id == void 0) && !sdIsString(val.id)) {
        throw new BoxSdkError({
            message: 'Expecting string for "id" of type "EnterpriseConfigurationV2025R0"',
        });
    }
    const id = val.id == void 0 ? void 0 : val.id;
    const type = val.type == void 0
        ? void 0
        : deserializeEnterpriseConfigurationV2025R0TypeField(val.type);
    const security = val.security == void 0
        ? void 0
        : deserializeEnterpriseConfigurationSecurityV2025R0(val.security);
    const contentAndSharing = val.content_and_sharing == void 0
        ? void 0
        : deserializeEnterpriseConfigurationContentAndSharingV2025R0(val.content_and_sharing);
    const userSettings = val.user_settings == void 0
        ? void 0
        : deserializeEnterpriseConfigurationUserSettingsV2025R0(val.user_settings);
    const shield = val.shield == void 0
        ? void 0
        : deserializeEnterpriseConfigurationShieldV2025R0(val.shield);
    return {
        id: id,
        type: type,
        security: security,
        contentAndSharing: contentAndSharing,
        userSettings: userSettings,
        shield: shield,
    };
}
