// box-node-sdk/lib-esm/schemas/v2025R0/enterpriseFeatureSettingV2025R0.js
import { serializeUserOrGroupReferenceV2025R0 } from './userOrGroupReferenceV2025R0.js';
import { deserializeUserOrGroupReferenceV2025R0 } from './userOrGroupReferenceV2025R0.js';
import { BoxSdkError } from '../../box/errors.js';
import { sdIsBoolean } from '../../serialization/json.js';
import { sdIsString } from '../../serialization/json.js';
import { sdIsList } from '../../serialization/json.js';
import { sdIsMap } from '../../serialization/json.js';
export function serializeEnterpriseFeatureSettingV2025R0FeatureField(val) {
    return { ['id']: val.id };
}
export function deserializeEnterpriseFeatureSettingV2025R0FeatureField(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "EnterpriseFeatureSettingV2025R0FeatureField"',
        });
    }
    if (!(val.id == void 0) && !sdIsString(val.id)) {
        throw new BoxSdkError({
            message: 'Expecting string for "id" of type "EnterpriseFeatureSettingV2025R0FeatureField"',
        });
    }
    const id = val.id == void 0 ? void 0 : val.id;
    return { id: id };
}
export function serializeEnterpriseFeatureSettingV2025R0(val) {
    return {
        ['id']: val.id,
        ['feature']: val.feature == void 0
            ? val.feature
            : serializeEnterpriseFeatureSettingV2025R0FeatureField(val.feature),
        ['state']: val.state,
        ['can_configure']: val.canConfigure,
        ['is_configured']: val.isConfigured,
        ['allowlist']: val.allowlist == void 0
            ? val.allowlist
            : val.allowlist.map(function (item) {
                return serializeUserOrGroupReferenceV2025R0(item);
            }),
        ['denylist']: val.denylist == void 0
            ? val.denylist
            : val.denylist.map(function (item) {
                return serializeUserOrGroupReferenceV2025R0(item);
            }),
    };
}
export function deserializeEnterpriseFeatureSettingV2025R0(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "EnterpriseFeatureSettingV2025R0"',
        });
    }
    if (!(val.id == void 0) && !sdIsString(val.id)) {
        throw new BoxSdkError({
            message: 'Expecting string for "id" of type "EnterpriseFeatureSettingV2025R0"',
        });
    }
    const id = val.id == void 0 ? void 0 : val.id;
    const feature = val.feature == void 0
        ? void 0
        : deserializeEnterpriseFeatureSettingV2025R0FeatureField(val.feature);
    if (!(val.state == void 0) && !sdIsString(val.state)) {
        throw new BoxSdkError({
            message: 'Expecting string for "state" of type "EnterpriseFeatureSettingV2025R0"',
        });
    }
    const state = val.state == void 0 ? void 0 : val.state;
    if (!(val.can_configure == void 0) && !sdIsBoolean(val.can_configure)) {
        throw new BoxSdkError({
            message: 'Expecting boolean for "can_configure" of type "EnterpriseFeatureSettingV2025R0"',
        });
    }
    const canConfigure = val.can_configure == void 0 ? void 0 : val.can_configure;
    if (!(val.is_configured == void 0) && !sdIsBoolean(val.is_configured)) {
        throw new BoxSdkError({
            message: 'Expecting boolean for "is_configured" of type "EnterpriseFeatureSettingV2025R0"',
        });
    }
    const isConfigured = val.is_configured == void 0 ? void 0 : val.is_configured;
    if (!(val.allowlist == void 0) && !sdIsList(val.allowlist)) {
        throw new BoxSdkError({
            message: 'Expecting array for "allowlist" of type "EnterpriseFeatureSettingV2025R0"',
        });
    }
    const allowlist = val.allowlist == void 0
        ? void 0
        : sdIsList(val.allowlist)
            ? val.allowlist.map(function (itm) {
                return deserializeUserOrGroupReferenceV2025R0(itm);
            })
            : [];
    if (!(val.denylist == void 0) && !sdIsList(val.denylist)) {
        throw new BoxSdkError({
            message: 'Expecting array for "denylist" of type "EnterpriseFeatureSettingV2025R0"',
        });
    }
    const denylist = val.denylist == void 0
        ? void 0
        : sdIsList(val.denylist)
            ? val.denylist.map(function (itm) {
                return deserializeUserOrGroupReferenceV2025R0(itm);
            })
            : [];
    return {
        id: id,
        feature: feature,
        state: state,
        canConfigure: canConfigure,
        isConfigured: isConfigured,
        allowlist: allowlist,
        denylist: denylist,
    };
}
