// box-node-sdk/lib-esm/schemas/v2025R0/sharedLinkPermissionsV2025R0.js
import { BoxSdkError } from '../../box/errors.js';
import { sdIsString } from '../../serialization/json.js';
import { sdIsMap } from '../../serialization/json.js';
export function serializeSharedLinkPermissionsV2025R0(val) {
    return {
        ['shared_links_option']: val.sharedLinksOption,
        ['default_shared_link_type']: val.defaultSharedLinkType,
        ['notes_shared_link_option']: val.notesSharedLinkOption,
        ['default_notes_shared_link_type']: val.defaultNotesSharedLinkType,
    };
}
export function deserializeSharedLinkPermissionsV2025R0(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "SharedLinkPermissionsV2025R0"',
        });
    }
    if (!(val.shared_links_option == void 0) &&
        !sdIsString(val.shared_links_option)) {
        throw new BoxSdkError({
            message: 'Expecting string for "shared_links_option" of type "SharedLinkPermissionsV2025R0"',
        });
    }
    const sharedLinksOption = val.shared_links_option == void 0 ? void 0 : val.shared_links_option;
    if (!(val.default_shared_link_type == void 0) &&
        !sdIsString(val.default_shared_link_type)) {
        throw new BoxSdkError({
            message: 'Expecting string for "default_shared_link_type" of type "SharedLinkPermissionsV2025R0"',
        });
    }
    const defaultSharedLinkType = val.default_shared_link_type == void 0
        ? void 0
        : val.default_shared_link_type;
    if (!(val.notes_shared_link_option == void 0) &&
        !sdIsString(val.notes_shared_link_option)) {
        throw new BoxSdkError({
            message: 'Expecting string for "notes_shared_link_option" of type "SharedLinkPermissionsV2025R0"',
        });
    }
    const notesSharedLinkOption = val.notes_shared_link_option == void 0
        ? void 0
        : val.notes_shared_link_option;
    if (!(val.default_notes_shared_link_type == void 0) &&
        !sdIsString(val.default_notes_shared_link_type)) {
        throw new BoxSdkError({
            message: 'Expecting string for "default_notes_shared_link_type" of type "SharedLinkPermissionsV2025R0"',
        });
    }
    const defaultNotesSharedLinkType = val.default_notes_shared_link_type == void 0
        ? void 0
        : val.default_notes_shared_link_type;
    return {
        sharedLinksOption: sharedLinksOption,
        defaultSharedLinkType: defaultSharedLinkType,
        notesSharedLinkOption: notesSharedLinkOption,
        defaultNotesSharedLinkType: defaultNotesSharedLinkType,
    };
}
