// box-node-sdk/lib-esm/schemas/v2025R0/userReferenceV2025R0.js
import { BoxSdkError } from '../../box/errors.js';
import { sdIsString } from '../../serialization/json.js';
import { sdIsMap } from '../../serialization/json.js';
export class UserReferenceV2025R0 {
    /**
     * The value is always `user`. */
    type = 'user';
    /**
     * The unique identifier for the user. */
    id;
    rawData;
    constructor(fields) {
        if (fields.type !== undefined) {
            this.type = fields.type;
        }
        if (fields.id !== undefined) {
            this.id = fields.id;
        }
        if (fields.rawData !== undefined) {
            this.rawData = fields.rawData;
        }
    }
}
export function serializeUserReferenceV2025R0TypeField(val) {
    return val;
}
export function deserializeUserReferenceV2025R0TypeField(val) {
    if (val == 'user') {
        return val;
    }
    throw new BoxSdkError({
        message: "Can't deserialize UserReferenceV2025R0TypeField",
    });
}
export function serializeUserReferenceV2025R0(val) {
    return {
        ['type']: serializeUserReferenceV2025R0TypeField(val.type),
        ['id']: val.id,
    };
}
export function deserializeUserReferenceV2025R0(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "UserReferenceV2025R0"',
        });
    }
    if (val.type == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "type" of type "UserReferenceV2025R0" to be defined',
        });
    }
    const type = deserializeUserReferenceV2025R0TypeField(val.type);
    if (val.id == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "id" of type "UserReferenceV2025R0" to be defined',
        });
    }
    if (!sdIsString(val.id)) {
        throw new BoxSdkError({
            message: 'Expecting string for "id" of type "UserReferenceV2025R0"',
        });
    }
    const id = val.id;
    return { type: type, id: id };
}
export function serializeUserReferenceV2025R0Input(val) {
    return {
        ['type']: val.type == void 0
            ? val.type
            : serializeUserReferenceV2025R0TypeField(val.type),
        ['id']: val.id,
    };
}
export function deserializeUserReferenceV2025R0Input(val) {
    if (!sdIsMap(val)) {
        throw new BoxSdkError({
            message: 'Expecting a map for "UserReferenceV2025R0Input"',
        });
    }
    const type = val.type == void 0
        ? void 0
        : deserializeUserReferenceV2025R0TypeField(val.type);
    if (val.id == void 0) {
        throw new BoxSdkError({
            message: 'Expecting "id" of type "UserReferenceV2025R0Input" to be defined',
        });
    }
    if (!sdIsString(val.id)) {
        throw new BoxSdkError({
            message: 'Expecting string for "id" of type "UserReferenceV2025R0Input"',
        });
    }
    const id = val.id;
    return { type: type, id: id };
}
