// box-node-sdk/lib/managers/externalUsers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExternalUsersManager = exports.SubmitJobToDeleteExternalUsersV2025R0Headers = exports.SubmitJobToDeleteExternalUsersV2025R0Optionals = void 0;
const externalUsersSubmitDeleteJobResponseV2025R0_1 = require("../schemas/v2025R0/externalUsersSubmitDeleteJobResponseV2025R0.js");
const externalUsersSubmitDeleteJobRequestV2025R0_1 = require("../schemas/v2025R0/externalUsersSubmitDeleteJobRequestV2025R0.js");
const network_1 = require("../networking/network.js");
const fetchOptions_1 = require("../networking/fetchOptions.js");
const utils_1 = require("../internal/utils.js");
const utils_2 = require("../internal/utils.js");
class SubmitJobToDeleteExternalUsersV2025R0Optionals {
    constructor(fields) {
        this.headers = new SubmitJobToDeleteExternalUsersV2025R0Headers({});
        this.cancellationToken = void 0;
        if (fields.headers !== undefined) {
            this.headers = fields.headers;
        }
        if (fields.cancellationToken !== undefined) {
            this.cancellationToken = fields.cancellationToken;
        }
    }
}
exports.SubmitJobToDeleteExternalUsersV2025R0Optionals = SubmitJobToDeleteExternalUsersV2025R0Optionals;
class SubmitJobToDeleteExternalUsersV2025R0Headers {
    constructor(fields) {
        /**
         * Version header. */
        this.boxVersion = '2025.0';
        /**
         * Extra headers that will be included in the HTTP request. */
        this.extraHeaders = {};
        if (fields.boxVersion !== undefined) {
            this.boxVersion = fields.boxVersion;
        }
        if (fields.extraHeaders !== undefined) {
            this.extraHeaders = fields.extraHeaders;
        }
    }
}
exports.SubmitJobToDeleteExternalUsersV2025R0Headers = SubmitJobToDeleteExternalUsersV2025R0Headers;
class ExternalUsersManager {
    constructor(fields) {
        this.networkSession = new network_1.NetworkSession({});
        if (fields.auth !== undefined) {
            this.auth = fields.auth;
        }
        if (fields.networkSession !== undefined) {
            this.networkSession = fields.networkSession;
        }
    }
    /**
     * Delete external users from current user enterprise. This will remove each
     * external user from all invited collaborations within the current enterprise.
     * @param {ExternalUsersSubmitDeleteJobRequestV2025R0} requestBody Request body of submitJobToDeleteExternalUsersV2025R0 method
     * @param {SubmitJobToDeleteExternalUsersV2025R0OptionalsInput} optionalsInput
     * @returns {Promise<ExternalUsersSubmitDeleteJobResponseV2025R0>}
     */
    submitJobToDeleteExternalUsersV2025R0(requestBody_1) {
        return __awaiter(this, arguments, void 0, function* (requestBody, optionalsInput = {}) {
            const optionals = new SubmitJobToDeleteExternalUsersV2025R0Optionals({
                headers: optionalsInput.headers,
                cancellationToken: optionalsInput.cancellationToken,
            });
            const headers = optionals.headers;
            const cancellationToken = optionals.cancellationToken;
            const headersMap = (0, utils_1.prepareParams)(Object.assign({ ['box-version']: (0, utils_2.toString)(headers.boxVersion) }, headers.extraHeaders));
            const response = yield this.networkSession.networkClient.fetch(new fetchOptions_1.FetchOptions({
                url: ''.concat(this.networkSession.baseUrls.baseUrl, '/2.0/external_users/submit_delete_job'),
                method: 'POST',
                headers: headersMap,
                data: (0, externalUsersSubmitDeleteJobRequestV2025R0_1.serializeExternalUsersSubmitDeleteJobRequestV2025R0)(requestBody),
                contentType: 'application/json',
                responseFormat: 'json',
                auth: this.auth,
                networkSession: this.networkSession,
                cancellationToken: cancellationToken,
            }));
            return Object.assign(Object.assign({}, (0, externalUsersSubmitDeleteJobResponseV2025R0_1.deserializeExternalUsersSubmitDeleteJobResponseV2025R0)(response.data)), { rawData: response.data });
        });
    }
}
exports.ExternalUsersManager = ExternalUsersManager;
