// box-node-sdk/lib/schemas/aiStudioAgentSpreadsheetTool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeAiStudioAgentSpreadsheetTool = serializeAiStudioAgentSpreadsheetTool;
exports.deserializeAiStudioAgentSpreadsheetTool = deserializeAiStudioAgentSpreadsheetTool;
const aiLlmEndpointParams_1 = require("./aiLlmEndpointParams.js");
const aiAgentSpreadsheetTool_1 = require("./aiAgentSpreadsheetTool.js");
const errors_1 = require("../box/errors.js");
const json_1 = require("../serialization/json.js");
const json_2 = require("../serialization/json.js");
const json_3 = require("../serialization/json.js");
function serializeAiStudioAgentSpreadsheetTool(val) {
    const base = (0, aiAgentSpreadsheetTool_1.serializeAiAgentSpreadsheetTool)(val);
    if (!(0, json_3.sdIsMap)(base)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting a map for "AiStudioAgentSpreadsheetTool"',
        });
    }
    return Object.assign(Object.assign({}, base), {});
}
function deserializeAiStudioAgentSpreadsheetTool(val) {
    if (!(0, json_3.sdIsMap)(val)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting a map for "AiStudioAgentSpreadsheetTool"',
        });
    }
    if (!(val.model == void 0) && !(0, json_2.sdIsString)(val.model)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting string for "model" of type "AiStudioAgentSpreadsheetTool"',
        });
    }
    const model = val.model == void 0 ? void 0 : val.model;
    if (!(val.num_tokens_for_completion == void 0) &&
        !(0, json_1.sdIsNumber)(val.num_tokens_for_completion)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting number for "num_tokens_for_completion" of type "AiStudioAgentSpreadsheetTool"',
        });
    }
    const numTokensForCompletion = val.num_tokens_for_completion == void 0
        ? void 0
        : val.num_tokens_for_completion;
    const llmEndpointParams = val.llm_endpoint_params == void 0
        ? void 0
        : (0, aiLlmEndpointParams_1.deserializeAiLlmEndpointParams)(val.llm_endpoint_params);
    return {
        model: model,
        numTokensForCompletion: numTokensForCompletion,
        llmEndpointParams: llmEndpointParams,
    };
}
