// box-node-sdk/lib/schemas/retentionPolicyAssignments.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeRetentionPolicyAssignments = serializeRetentionPolicyAssignments;
exports.deserializeRetentionPolicyAssignments = deserializeRetentionPolicyAssignments;
const retentionPolicyAssignment_1 = require("./retentionPolicyAssignment.js");
const retentionPolicyAssignment_2 = require("./retentionPolicyAssignment.js");
const errors_1 = require("../box/errors.js");
const json_1 = require("../serialization/json.js");
const json_2 = require("../serialization/json.js");
const json_3 = require("../serialization/json.js");
const json_4 = require("../serialization/json.js");
function serializeRetentionPolicyAssignments(val) {
    return {
        ['entries']: val.entries == void 0
            ? val.entries
            : val.entries.map(function (item) {
                return (0, retentionPolicyAssignment_1.serializeRetentionPolicyAssignment)(item);
            }),
        ['limit']: val.limit,
        ['next_marker']: val.nextMarker,
    };
}
function deserializeRetentionPolicyAssignments(val) {
    if (!(0, json_4.sdIsMap)(val)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting a map for "RetentionPolicyAssignments"',
        });
    }
    if (!(val.entries == void 0) && !(0, json_3.sdIsList)(val.entries)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting array for "entries" of type "RetentionPolicyAssignments"',
        });
    }
    const entries = val.entries == void 0
        ? void 0
        : (0, json_3.sdIsList)(val.entries)
            ? val.entries.map(function (itm) {
                return (0, retentionPolicyAssignment_2.deserializeRetentionPolicyAssignment)(itm);
            })
            : [];
    if (!(val.limit == void 0) && !(0, json_1.sdIsNumber)(val.limit)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting number for "limit" of type "RetentionPolicyAssignments"',
        });
    }
    const limit = val.limit == void 0 ? void 0 : val.limit;
    if (!(val.next_marker == void 0) && !(0, json_2.sdIsString)(val.next_marker)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting string for "next_marker" of type "RetentionPolicyAssignments"',
        });
    }
    const nextMarker = val.next_marker == void 0 ? void 0 : val.next_marker;
    return {
        entries: entries,
        limit: limit,
        nextMarker: nextMarker,
    };
}
