// box-node-sdk/lib/schemas/signRequestSignerInputDateEuValidation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeSignRequestSignerInputDateEuValidationValidationTypeField = serializeSignRequestSignerInputDateEuValidationValidationTypeField;
exports.deserializeSignRequestSignerInputDateEuValidationValidationTypeField = deserializeSignRequestSignerInputDateEuValidationValidationTypeField;
exports.serializeSignRequestSignerInputDateEuValidation = serializeSignRequestSignerInputDateEuValidation;
exports.deserializeSignRequestSignerInputDateEuValidation = deserializeSignRequestSignerInputDateEuValidation;
const errors_1 = require("../box/errors.js");
const json_1 = require("../serialization/json.js");
const json_2 = require("../serialization/json.js");
function serializeSignRequestSignerInputDateEuValidationValidationTypeField(val) {
    return val;
}
function deserializeSignRequestSignerInputDateEuValidationValidationTypeField(val) {
    if (val == 'date_eu') {
        return val;
    }
    if ((0, json_1.sdIsString)(val)) {
        return val;
    }
    throw new errors_1.BoxSdkError({
        message: "Can't deserialize SignRequestSignerInputDateEuValidationValidationTypeField",
    });
}
function serializeSignRequestSignerInputDateEuValidation(val) {
    return {
        ['validation_type']: val.validationType == void 0
            ? val.validationType
            : serializeSignRequestSignerInputDateEuValidationValidationTypeField(val.validationType),
    };
}
function deserializeSignRequestSignerInputDateEuValidation(val) {
    if (!(0, json_2.sdIsMap)(val)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting a map for "SignRequestSignerInputDateEuValidation"',
        });
    }
    const validationType = val.validation_type == void 0
        ? void 0
        : deserializeSignRequestSignerInputDateEuValidationValidationTypeField(val.validation_type);
    return {
        validationType: validationType,
    };
}
