// box-node-sdk/lib/schemas/storagePolicyAssignments.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeStoragePolicyAssignments = serializeStoragePolicyAssignments;
exports.deserializeStoragePolicyAssignments = deserializeStoragePolicyAssignments;
const storagePolicyAssignment_1 = require("./storagePolicyAssignment.js");
const storagePolicyAssignment_2 = require("./storagePolicyAssignment.js");
const errors_1 = require("../box/errors.js");
const json_1 = require("../serialization/json.js");
const json_2 = require("../serialization/json.js");
const json_3 = require("../serialization/json.js");
const json_4 = require("../serialization/json.js");
function serializeStoragePolicyAssignments(val) {
    return {
        ['limit']: val.limit,
        ['next_marker']: val.nextMarker,
        ['prev_marker']: val.prevMarker,
        ['entries']: val.entries == void 0
            ? val.entries
            : val.entries.map(function (item) {
                return (0, storagePolicyAssignment_1.serializeStoragePolicyAssignment)(item);
            }),
    };
}
function deserializeStoragePolicyAssignments(val) {
    if (!(0, json_4.sdIsMap)(val)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting a map for "StoragePolicyAssignments"',
        });
    }
    if (!(val.limit == void 0) && !(0, json_1.sdIsNumber)(val.limit)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting number for "limit" of type "StoragePolicyAssignments"',
        });
    }
    const limit = val.limit == void 0 ? void 0 : val.limit;
    if (!(val.next_marker == void 0) && !(0, json_2.sdIsString)(val.next_marker)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting string for "next_marker" of type "StoragePolicyAssignments"',
        });
    }
    const nextMarker = val.next_marker == void 0 ? void 0 : val.next_marker;
    if (!(val.prev_marker == void 0) && !(0, json_2.sdIsString)(val.prev_marker)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting string for "prev_marker" of type "StoragePolicyAssignments"',
        });
    }
    const prevMarker = val.prev_marker == void 0 ? void 0 : val.prev_marker;
    if (!(val.entries == void 0) && !(0, json_3.sdIsList)(val.entries)) {
        throw new errors_1.BoxSdkError({
            message: 'Expecting array for "entries" of type "StoragePolicyAssignments"',
        });
    }
    const entries = val.entries == void 0
        ? void 0
        : (0, json_3.sdIsList)(val.entries)
            ? val.entries.map(function (itm) {
                return (0, storagePolicyAssignment_2.deserializeStoragePolicyAssignment)(itm);
            })
            : [];
    return {
        limit: limit,
        nextMarker: nextMarker,
        prevMarker: prevMarker,
        entries: entries,
    };
}
